File: ruby/rails/lib/postgres_petstore_repository.rb
require 'pg'
require 'json'
require_relative 'petstore_errors'

# PostgreSQL-backed implementation of the Petstore repository.
# Uses raw SQL via the pg gem — no ActiveRecord.
class PostgresPetstoreRepository
  def initialize(options = {})
    @options = {
      host:     ENV.fetch('POSTGRES_HOST', 'localhost'),
      port:     ENV.fetch('POSTGRES_PORT', '5434').to_i,
      dbname:   ENV.fetch('POSTGRES_DB', 'ruby-rails'),
      user:     ENV.fetch('POSTGRES_USER', 'myuser'),
      password: ENV.fetch('POSTGRES_PASSWORD', 'mypassword')
    }.merge(options)
  end

  # ---------------------------------------------------------------------------
  # Pet operations
  # ---------------------------------------------------------------------------

  def add_pet(pet)
    id = fetch(pet, :id, 'id')&.to_i || next_id('pet')
    exec_params(
      <<~SQL,
        INSERT INTO pet (id, name, category, photo_urls, tags, status)
        VALUES ($1, $2, $3, cast($4 as json), cast($5 as json), cast($6 as pet_status))
        ON CONFLICT (id) DO UPDATE SET
          name = EXCLUDED.name, category = EXCLUDED.category,
          photo_urls = EXCLUDED.photo_urls, tags = EXCLUDED.tags,
          status = EXCLUDED.status
      SQL
      [id, pet_name(pet), pet_category(pet),
       JSON.dump(pet_photo_urls(pet)), json_or_nil(pet_tags(pet)), pet_status(pet)]
    )
    map_pet_from_input(pet, id)
  end

  def update_pet(pet)
    id = fetch(pet, :id, 'id')&.to_i
    raise InvalidInputError, 'Pet ID required for update' if id.nil?

    result = exec_params(
      <<~SQL,
        INSERT INTO pet (id, name, category, photo_urls, tags, status)
        VALUES ($1, $2, $3, cast($4 as json), cast($5 as json), cast($6 as pet_status))
        ON CONFLICT (id) DO UPDATE SET
          name = EXCLUDED.name, category = EXCLUDED.category,
          photo_urls = EXCLUDED.photo_urls, tags = EXCLUDED.tags,
          status = EXCLUDED.status
        RETURNING id
      SQL
      [id, pet_name(pet), pet_category(pet),
       JSON.dump(pet_photo_urls(pet)), json_or_nil(pet_tags(pet)), pet_status(pet)]
    )
    raise NotFoundError, 'Pet not found' if result.ntuples.zero?
    map_pet_from_input(pet, id)
  end

  def find_pet_by_id(pet_id)
    result = exec_params(
      'SELECT id, name, category, photo_urls, tags, status::text FROM pet WHERE id = $1',
      [pet_id.to_i]
    )
    raise NotFoundError, 'Pet not found' if result.ntuples.zero?
    map_pet_row(result[0])
  end

  def delete_pet(pet_id)
    exec_params('DELETE FROM pet WHERE id = $1', [pet_id.to_i])
  end

  def find_pets_by_status(status)
    result = exec_params(
      'SELECT id, name, category, photo_urls, tags, status::text FROM pet WHERE status = cast($1 as pet_status)',
      [status]
    )
    result.map { |row| map_pet_row(row) }
  end

  def find_pets_by_tags(tags)
    return [] if tags.nil? || tags.empty?

    conditions = tags.each_with_index.map { |_, i| "tags::jsonb @> cast($#{i + 1} as jsonb)" }.join(' OR ')
    params = tags.map { |t| JSON.dump([{ 'name' => t }]) }
    result = exec_params(
      "SELECT id, name, category, photo_urls, tags, status::text FROM pet WHERE #{conditions}",
      params
    )
    result.map { |row| map_pet_row(row) }
  end

  def update_pet_with_form(pet_id, name, status)
    sets = []
    params = []
    if name
      sets << "name = $#{params.size + 1}"
      params << name
    end
    if status
      sets << "status = cast($#{params.size + 1} as pet_status)"
      params << status
    end
    return if sets.empty?

    params << pet_id.to_i
    exec_params("UPDATE pet SET #{sets.join(', ')} WHERE id = $#{params.size}", params)
  end

  def upload_file(pet_id, content, content_type, metadata)
    exists = exec_params('SELECT 1 FROM pet WHERE id = $1', [pet_id.to_i])
    raise NotFoundError, 'Pet not found' if exists.ntuples.zero?

    encoded = PG::Connection.escape_bytea(content.b)
    exec_params(
      <<~SQL,
        INSERT INTO pet_photo (id, pet_id, content_type, metadata, content)
        VALUES ((SELECT COALESCE(MAX(id), 0) + 1 FROM pet_photo), $1, $2, $3, $4::bytea)
      SQL
      [pet_id.to_i, content_type, metadata, encoded]
    )
    content.bytesize
  end

  # ---------------------------------------------------------------------------
  # Store operations
  # ---------------------------------------------------------------------------

  def get_inventory
    result = exec_params(
      "SELECT status::text, cast(COUNT(*) as int) AS cnt FROM pet GROUP BY status",
      []
    )
    result.each_with_object({}) do |row, inv|
      inv[row['status']] = row['cnt'].to_i if row['status']
    end
  end

  def place_order(order)
    id = fetch(order, :id, 'id')&.to_i || next_id('"order"')
    exec_params(
      <<~SQL,
        INSERT INTO "order" (id, pet_id, quantity, ship_date, status, complete)
        VALUES ($1, $2, $3, $4, cast($5 as order_status), $6)
        ON CONFLICT (id) DO UPDATE SET
          pet_id = EXCLUDED.pet_id, quantity = EXCLUDED.quantity,
          ship_date = EXCLUDED.ship_date, status = EXCLUDED.status,
          complete = EXCLUDED.complete
      SQL
      [id, order_pet_id(order), order_quantity(order),
       order_ship_date(order), order_status(order), order_complete(order)]
    )
    map_order_from_input(order, id)
  end

  def find_order_by_id(order_id)
    result = exec_params(
      'SELECT id, pet_id, quantity, ship_date, status::text, complete FROM "order" WHERE id = $1',
      [order_id.to_i]
    )
    raise NotFoundError, 'Order not found' if result.ntuples.zero?
    map_order_row(result[0])
  end

  def delete_order(order_id)
    exec_params('DELETE FROM "order" WHERE id = $1', [order_id.to_i])
  end

  # ---------------------------------------------------------------------------
  # User operations
  # ---------------------------------------------------------------------------

  def create_user(user)
    exec_params(
      <<~SQL,
        INSERT INTO "user" (id, username, first_name, last_name, email, password, phone, user_status)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
        ON CONFLICT (username) DO UPDATE SET
          id = EXCLUDED.id, first_name = EXCLUDED.first_name,
          last_name = EXCLUDED.last_name, email = EXCLUDED.email,
          password = EXCLUDED.password, phone = EXCLUDED.phone,
          user_status = EXCLUDED.user_status
      SQL
      [user_id(user), user_username(user), user_first_name(user), user_last_name(user),
       user_email(user), user_password(user), user_phone(user), user_status_val(user)]
    )
    user
  end

  def create_users_with_list(users)
    users.each { |u| create_user(u) }
  end

  def find_user_by_username(username)
    result = exec_params(
      'SELECT id, username, first_name, last_name, email, password, phone, user_status FROM "user" WHERE username = $1',
      [username]
    )
    raise NotFoundError, 'User not found' if result.ntuples.zero?
    map_user_row(result[0])
  end

  def update_user(username, user)
    result = exec_params(
      <<~SQL,
        UPDATE "user" SET
          id = $1, first_name = $2, last_name = $3, email = $4,
          password = $5, phone = $6, user_status = $7
        WHERE username = $8
        RETURNING username
      SQL
      [user_id(user), user_first_name(user), user_last_name(user),
       user_email(user), user_password(user), user_phone(user),
       user_status_val(user), username]
    )
    raise NotFoundError, 'User not found' if result.ntuples.zero?
  end

  def delete_user(username)
    exec_params('DELETE FROM "user" WHERE username = $1', [username])
  end

  def login_user(username, password)
    result = exec_params(
      'SELECT 1 FROM "user" WHERE username = $1 AND password = $2',
      [username, password]
    )
    result.ntuples.positive?
  end

  private

  def conn
    Thread.current[:petstore_pg_conn] ||= PG.connect(@options)
  end

  def exec_params(sql, params)
    conn.exec_params(sql, params)
  end

  def next_id(table)
    result = exec_params("SELECT COALESCE(MAX(id), 0) + 1 AS nid FROM #{table}", [])
    result[0]['nid'].to_i
  end

  # Pet helpers
  def fetch(obj, sym_key, str_key) = obj[sym_key] || obj[str_key]
  def pet_name(pet)        = fetch(pet, :name, 'name')
  def pet_status(pet)      = fetch(pet, :status, 'status')
  def pet_photo_urls(pet)  = fetch(pet, :photoUrls, 'photoUrls') || fetch(pet, :photo_urls, 'photo_urls') || []
  def pet_tags(pet)        = fetch(pet, :tags, 'tags')

  def pet_category(pet)
    cat = fetch(pet, :category, 'category')
    cat ? JSON.dump(cat) : nil
  end

  def json_or_nil(val) = val ? JSON.dump(val) : nil

  def map_pet_from_input(pet, id)
    {
      'id'        => id,
      'name'      => pet_name(pet),
      'category'  => fetch(pet, :category, 'category'),
      'photoUrls' => pet_photo_urls(pet),
      'tags'      => pet_tags(pet),
      'status'    => pet_status(pet)
    }.compact
  end

  def map_pet_row(row)
    {
      'id'        => row['id']&.to_i,
      'name'      => row['name'],
      'category'  => row['category'] ? JSON.parse(row['category']) : nil,
      'photoUrls' => row['photo_urls'] ? JSON.parse(row['photo_urls']) : [],
      'tags'      => row['tags'] ? JSON.parse(row['tags']) : nil,
      'status'    => row['status']
    }.compact
  end

  # Order helpers
  def order_pet_id(o)    = (v = fetch(o, :petId, 'petId') || fetch(o, :pet_id, 'pet_id')) ? v.to_i : nil
  def order_quantity(o)  = (v = fetch(o, :quantity, 'quantity')) ? v.to_i : nil
  def order_status(o)    = fetch(o, :status, 'status')
  def order_complete(o)  = fetch(o, :complete, 'complete')

  def order_ship_date(order)
    v = fetch(order, :shipDate, 'shipDate') || fetch(order, :ship_date, 'ship_date')
    v ? Time.parse(v.to_s) : nil
  end

  def map_order_from_input(order, id)
    {
      'id'       => id,
      'petId'    => order_pet_id(order),
      'quantity' => order_quantity(order),
      'shipDate' => (sd = order_ship_date(order)) ? sd.iso8601 : nil,
      'status'   => order_status(order),
      'complete' => order_complete(order)
    }.compact
  end

  def map_order_row(row)
    {
      'id'       => row['id']&.to_i,
      'petId'    => row['pet_id']&.to_i,
      'quantity' => row['quantity']&.to_i,
      'shipDate' => row['ship_date'],
      'status'   => row['status'],
      'complete' => row['complete'] == 't'
    }.compact
  end

  # User helpers
  def user_id(u)         = (v = fetch(u, :id, 'id')) ? v.to_i : nil
  def user_username(u)   = fetch(u, :username, 'username')
  def user_first_name(u) = fetch(u, :firstName, 'firstName') || fetch(u, :first_name, 'first_name')
  def user_last_name(u)  = fetch(u, :lastName, 'lastName') || fetch(u, :last_name, 'last_name')
  def user_email(u)      = fetch(u, :email, 'email')
  def user_password(u)   = fetch(u, :password, 'password')
  def user_phone(u)      = fetch(u, :phone, 'phone')
  def user_status_val(u) = (v = fetch(u, :userStatus, 'userStatus') || fetch(u, :user_status, 'user_status')) ? v.to_i : nil

  def map_user_row(row)
    {
      'id'         => row['id']&.to_i,
      'username'   => row['username'],
      'firstName'  => row['first_name'],
      'lastName'   => row['last_name'],
      'email'      => row['email'],
      'password'   => row['password'],
      'phone'      => row['phone'],
      'userStatus' => row['user_status']&.to_i
    }.compact
  end
end

File: petshop-stacks/kotlin/ktor/src/main/kotlin/com/example/petstore/repository/PostgresPetstoreRepository.kt
package com.example.petstore.repository

import com.example.petstore.models.*
import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.postgresql.util.PGobject
import java.sql.Connection
import java.sql.ResultSet
import javax.sql.DataSource

/**
 * JDBC-backed repository. All public methods are suspend functions that shift
 * the blocking JDBC work onto [Dispatchers.IO], so Ktor's Netty event-loop
 * threads are never pinned during database I/O. Each method's JDBC work runs
 * entirely within a single withContext(Dispatchers.IO) block.
 */
class PostgresPetstoreRepository(dataSource: DataSource? = null) : PetstoreRepository {

    private val ds: DataSource = dataSource ?: createDataSource()

    private fun conn(): Connection = ds.connection

    // -------------------------------------------------------------------------
    // Pet
    // -------------------------------------------------------------------------

    override suspend fun addPet(pet: Pet): Pet = withContext(Dispatchers.IO) {
        conn().use { c ->
            val id = pet.id ?: nextId(c, "pet")
            val saved = pet.copy(id = id)
            upsertPet(c, saved)
            fetchPetById(c, id)!!
        }
    }

    override suspend fun updatePet(pet: Pet): Pet {
        val id = pet.id ?: throw InvalidInputException("Pet id required for update")
        return withContext(Dispatchers.IO) {
            conn().use { c ->
                val existing = fetchPetById(c, id) ?: throw NotFoundException("Pet not found: $id")
                upsertPet(c, existing.copy(
                    name = pet.name,
                    photoUrls = pet.photoUrls,
                    category = pet.category,
                    tags = pet.tags,
                    status = pet.status
                ))
                fetchPetById(c, id)!!
            }
        }
    }

    override suspend fun findPetsByStatus(status: String): List<Pet> = withContext(Dispatchers.IO) {
        conn().use { c ->
            c.prepareStatement(
                "SELECT id, name, photo_urls, category, tags, status::text FROM pet WHERE status::text = ?"
            ).use { ps ->
                ps.setString(1, status)
                ps.executeQuery().use { rs -> rs.toPetList() }
            }
        }
    }

    override suspend fun findPetsByTags(tags: List<String>): List<Pet> = withContext(Dispatchers.IO) {
        conn().use { c ->
            c.prepareStatement(
                "SELECT id, name, photo_urls, category, tags, status::text FROM pet"
            ).use { ps ->
                val allPets = ps.executeQuery().use { rs -> rs.toPetList() }
                allPets.filter { pet ->
                    pet.tags?.any { tag -> tags.contains(tag.name) } == true
                }
            }
        }
    }

    override suspend fun getPetById(petId: Long): Pet? = withContext(Dispatchers.IO) {
        fetchPetById(petId)
    }

    override suspend fun updatePetWithForm(petId: Long, name: String?, status: String?): Pet =
        withContext(Dispatchers.IO) {
            conn().use { c ->
                val existing = fetchPetById(c, petId) ?: throw NotFoundException("Pet not found: $petId")
                val updated = existing.copy(
                    name = name ?: existing.name,
                    status = status?.let { s -> Pet.Status.entries.find { it.value == s } } ?: existing.status
                )
                upsertPet(c, updated)
                fetchPetById(c, petId)!!
            }
        }

    override suspend fun deletePet(petId: Long): Unit = withContext(Dispatchers.IO) {
        conn().use { c ->
            fetchPetById(c, petId) ?: throw NotFoundException("Pet not found: $petId")
            c.prepareStatement("DELETE FROM pet WHERE id = ?").use { ps ->
                ps.setLong(1, petId)
                ps.executeUpdate()
            }
        }
    }

    override suspend fun uploadFile(petId: Long, additionalMetadata: String?, bytes: ByteArray): ModelApiResponse =
        withContext(Dispatchers.IO) {
            conn().use { c ->
                fetchPetById(c, petId) ?: throw NotFoundException("Pet not found: $petId")
                val photoId = nextId(c, "pet_photo")
                c.prepareStatement(
                    "INSERT INTO pet_photo (id, pet_id, metadata, content) VALUES (?, ?, ?, ?) " +
                        "ON CONFLICT (id) DO UPDATE SET content = excluded.content"
                ).use { ps ->
                    ps.setLong(1, photoId)
                    ps.setLong(2, petId)
                    ps.setString(3, additionalMetadata)
                    ps.setBytes(4, bytes)
                    ps.executeUpdate()
                }
            }
            ModelApiResponse(code = 200, type = "unknown", message = "File uploaded: ${bytes.size} bytes")
        }

    private fun fetchPetById(petId: Long): Pet? = conn().use { fetchPetById(it, petId) }

    /** Uses an existing connection so composite operations never hold one pool slot while acquiring another. */
    private fun fetchPetById(c: Connection, petId: Long): Pet? {
        c.prepareStatement(
            "SELECT id, name, photo_urls, category, tags, status::text FROM pet WHERE id = ?"
        ).use { ps ->
            ps.setLong(1, petId)
            return ps.executeQuery().use { rs -> if (rs.next()) rs.toPet() else null }
        }
    }

    private fun upsertPet(c: Connection, pet: Pet) {
        c.prepareStatement(
            """INSERT INTO pet (id, name, photo_urls, category, tags, status)
               VALUES (?, ?, ?, ?, ?, ?::pet_status)
               ON CONFLICT (id) DO UPDATE SET
                 name = excluded.name,
                 photo_urls = excluded.photo_urls,
                 category = excluded.category,
                 tags = excluded.tags,
                 status = excluded.status"""
        ).use { ps ->
            ps.setLong(1, pet.id!!)
            ps.setString(2, pet.name)
            ps.setObject(3, jsonObj(Json.encodeToString(pet.photoUrls)))
            ps.setString(4, pet.category?.let { Json.encodeToString(it) })
            ps.setObject(5, jsonObj(Json.encodeToString(pet.tags ?: emptyList<Tag>())))
            ps.setString(6, pet.status?.value)
            ps.executeUpdate()
        }
    }

    // -------------------------------------------------------------------------
    // Store
    // -------------------------------------------------------------------------

    override suspend fun getInventory(): Map<String, Int> = withContext(Dispatchers.IO) {
        conn().use { c ->
            c.prepareStatement(
                "SELECT status::text, COUNT(*) AS cnt FROM pet GROUP BY status"
            ).use { ps ->
                val result = mutableMapOf<String, Int>()
                ps.executeQuery().use { rs ->
                    while (rs.next()) {
                        result[rs.getString("status")] = rs.getInt("cnt")
                    }
                }
                result
            }
        }
    }

    override suspend fun placeOrder(order: Order): Order = withContext(Dispatchers.IO) {
        conn().use { c ->
            val id = order.id ?: nextId(c, "\"order\"")
            val saved = order.copy(id = id)
            upsertOrder(c, saved)
            fetchOrderById(c, id)!!
        }
    }

    override suspend fun getOrderById(orderId: Long): Order? = withContext(Dispatchers.IO) {
        fetchOrderById(orderId)
    }

    override suspend fun deleteOrder(orderId: Long): Unit = withContext(Dispatchers.IO) {
        conn().use { c ->
            fetchOrderById(c, orderId) ?: throw NotFoundException("Order not found: $orderId")
            c.prepareStatement("DELETE FROM \"order\" WHERE id = ?").use { ps ->
                ps.setLong(1, orderId)
                ps.executeUpdate()
            }
        }
    }

    private fun fetchOrderById(orderId: Long): Order? = conn().use { fetchOrderById(it, orderId) }

    private fun fetchOrderById(c: Connection, orderId: Long): Order? {
        c.prepareStatement(
            "SELECT id, pet_id, quantity, ship_date, status::text, complete FROM \"order\" WHERE id = ?"
        ).use { ps ->
            ps.setLong(1, orderId)
            return ps.executeQuery().use { rs -> if (rs.next()) rs.toOrder() else null }
        }
    }

    private fun upsertOrder(c: Connection, order: Order) {
        c.prepareStatement(
            """INSERT INTO "order" (id, pet_id, quantity, ship_date, status, complete)
               VALUES (?, ?, ?, ?::timestamp, ?::order_status, ?)
               ON CONFLICT (id) DO UPDATE SET
                 pet_id = excluded.pet_id,
                 quantity = excluded.quantity,
                 ship_date = excluded.ship_date,
                 status = excluded.status,
                 complete = excluded.complete"""
        ).use { ps ->
            ps.setLong(1, order.id!!)
            order.petId?.let { ps.setLong(2, it) } ?: ps.setNull(2, java.sql.Types.BIGINT)
            order.quantity?.let { ps.setInt(3, it) } ?: ps.setNull(3, java.sql.Types.INTEGER)
            ps.setString(4, order.shipDate)
            ps.setString(5, order.status?.value)
            order.complete?.let { ps.setBoolean(6, it) } ?: ps.setNull(6, java.sql.Types.BOOLEAN)
            ps.executeUpdate()
        }
    }

    // -------------------------------------------------------------------------
    // User
    // -------------------------------------------------------------------------

    override suspend fun createUser(user: User): User = withContext(Dispatchers.IO) {
        conn().use { c ->
            val id = user.id ?: nextId(c, "\"user\"")
            upsertUser(c, user.copy(id = id))
            fetchUserByName(c, user.username ?: "")!!
        }
    }

    override suspend fun createUsersWithList(users: List<User>): User = withContext(Dispatchers.IO) {
        conn().use { c ->
            for (user in users) {
                val id = user.id ?: nextId(c, "\"user\"")
                upsertUser(c, user.copy(id = id))
            }
            fetchUserByName(c, users.first().username ?: "")!!
        }
    }

    override suspend fun loginUser(username: String, password: String): String = withContext(Dispatchers.IO) {
        fetchUserByName(username) ?: throw NotFoundException("User not found: $username")
        "logged-in-$username"
    }

    override suspend fun logoutUser() { /* stateless no-op */ }

    override suspend fun getUserByName(username: String): User? = withContext(Dispatchers.IO) {
        fetchUserByName(username)
    }

    override suspend fun updateUser(username: String, user: User): Unit = withContext(Dispatchers.IO) {
        conn().use { c ->
            fetchUserByName(c, username) ?: throw NotFoundException("User not found: $username")
            upsertUser(c, user.copy(username = username))
        }
    }

    override suspend fun deleteUser(username: String): Unit = withContext(Dispatchers.IO) {
        conn().use { c ->
            fetchUserByName(c, username) ?: throw NotFoundException("User not found: $username")
            c.prepareStatement("DELETE FROM \"user\" WHERE username = ?").use { ps ->
                ps.setString(1, username)
                ps.executeUpdate()
            }
        }
    }

    private fun fetchUserByName(username: String): User? = conn().use { fetchUserByName(it, username) }

    private fun fetchUserByName(c: Connection, username: String): User? {
        c.prepareStatement(
            "SELECT id, username, first_name, last_name, email, password, phone, user_status FROM \"user\" WHERE username = ?"
        ).use { ps ->
            ps.setString(1, username)
            return ps.executeQuery().use { rs -> if (rs.next()) rs.toUser() else null }
        }
    }

    private fun upsertUser(c: Connection, user: User) {
        val uname = user.username ?: throw InvalidInputException("Username required")
        c.prepareStatement(
            """INSERT INTO "user" (id, username, first_name, last_name, email, password, phone, user_status)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?)
               ON CONFLICT (username) DO UPDATE SET
                 id = excluded.id,
                 first_name = excluded.first_name,
                 last_name = excluded.last_name,
                 email = excluded.email,
                 password = excluded.password,
                 phone = excluded.phone,
                 user_status = excluded.user_status"""
        ).use { ps ->
            user.id?.let { ps.setLong(1, it) } ?: ps.setNull(1, java.sql.Types.BIGINT)
            ps.setString(2, uname)
            ps.setString(3, user.firstName)
            ps.setString(4, user.lastName)
            ps.setString(5, user.email)
            ps.setString(6, user.password)
            ps.setString(7, user.phone)
            user.userStatus?.let { ps.setInt(8, it) } ?: ps.setNull(8, java.sql.Types.INTEGER)
            ps.executeUpdate()
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private fun nextId(c: Connection, table: String): Long {
        val seqName = "${table.replace("\"", "")}_id_seq"
        c.prepareStatement("SELECT nextval('$seqName')").use { ps ->
            return ps.executeQuery().use { rs ->
                rs.next()
                rs.getLong(1)
            }
        }
    }

    private fun jsonObj(value: String): PGobject {
        val obj = PGobject()
        obj.type = "json"
        obj.value = value
        return obj
    }

    private fun ResultSet.toPet(): Pet {
        val categoryJson = getString("category")
        val tagsJson = getString("tags")
        val photoUrlsJson = getString("photo_urls")
        return Pet(
            id = getLong("id"),
            name = getString("name"),
            photoUrls = if (photoUrlsJson != null) Json.decodeFromString(photoUrlsJson) else emptyList(),
            category = if (categoryJson != null) Json.decodeFromString(categoryJson) else null,
            tags = if (tagsJson != null) Json.decodeFromString(tagsJson) else null,
            status = getString("status")?.let { s -> Pet.Status.entries.find { it.value == s } }
        )
    }

    private fun ResultSet.toPetList(): List<Pet> {
        val list = mutableListOf<Pet>()
        while (next()) list.add(toPet())
        return list
    }

    private fun ResultSet.toOrder(): Order {
        return Order(
            id = getLong("id"),
            petId = getLong("pet_id").takeIf { it != 0L },
            quantity = getInt("quantity").takeIf { !wasNull() },
            shipDate = getString("ship_date"),
            status = getString("status")?.let { s -> Order.Status.entries.find { it.value == s } },
            complete = getBoolean("complete").takeIf { !wasNull() }
        )
    }

    private fun ResultSet.toUser(): User {
        return User(
            id = getLong("id").takeIf { it != 0L },
            username = getString("username"),
            firstName = getString("first_name"),
            lastName = getString("last_name"),
            email = getString("email"),
            password = getString("password"),
            phone = getString("phone"),
            userStatus = getInt("user_status").takeIf { !wasNull() }
        )
    }
}

// -------------------------------------------------------------------------
// Domain exceptions
// -------------------------------------------------------------------------

class NotFoundException(message: String) : RuntimeException(message)
class InvalidInputException(message: String) : RuntimeException(message)

// -------------------------------------------------------------------------
// DataSource factory
// -------------------------------------------------------------------------

fun createDataSource(): HikariDataSource {
    val host = System.getenv("POSTGRES_HOST") ?: "localhost"
    val port = System.getenv("POSTGRES_PORT")?.toInt() ?: 5434
    val db = System.getenv("POSTGRES_DB") ?: "kotlin-ktor"
    val user = System.getenv("POSTGRES_USER") ?: "myuser"
    val pass = System.getenv("POSTGRES_PASSWORD") ?: "mypassword"
    val url = System.getenv("DATABASE_URL") ?: "jdbc:postgresql://$host:$port/$db"

    return HikariDataSource(HikariConfig().apply {
        jdbcUrl = url
        username = user
        password = pass
        driverClassName = "org.postgresql.Driver"
        maximumPoolSize = System.getenv("HIKARI_MAXIMUM_POOL_SIZE")?.toInt() ?: 10
        System.getenv("HIKARI_MINIMUM_IDLE")?.let { minimumIdle = it.toInt() }
    })
}
